core/src/main/java/hudson/model/AbstractItem.java

```diff
@@ -228,6 +233,102 @@ protected void doSetName(String name) {
         this.name = name;
     }
 
+    /**
+     * @return whether {@link #name} can be modified by a user
+     */
+    public boolean isNameEditable() {
+        return false;
+    }
+
+    /**
+     * Renames this item
+     */
+    @RequirePOST
+    @Restricted(NoExternalUse.class)
+    public void doDoRename2(StaplerRequest req, StaplerResponse rsp) throws IOException {
+        String newName = req.getParameter("newName");
+        newName = newName == null ? null : newName.trim();
+        FormValidation validationError = doCheckNewName(newName);
+        if (validationError.kind != FormValidation.Kind.OK) {
+            throw new Failure(validationError.getMessage());
+        }
+
+        renameTo(newName);
+        // send to the new job page
+        // note we can't use getUrl() because that would pick up old name in the
+        // Ancestor.getUrl()
+        rsp.sendRedirect2("../" + newName);
+    }
+
+    /**
+     * Called by {@link #doDoRename2} and {@code rename.jelly} to validate renames.
+     * @return {@link FormValidation#ok} if this item can be renamed as specified, otherwise
+     * {@link FormValidation#error} with a message explaining the problem.
+     */
+    @Restricted(NoExternalUse.class)
+    public @Nonnull FormValidation doCheckNewName(@QueryParameter String newName) {
+        // TODO: Create an Item.RENAME permission to use here, see JENKINS-18649.
+        if (!hasPermission(Item.CONFIGURE)) {
+            checkPermission(Item.CREATE);
```

Comment:
as noted in JENKINS-18649 this is actually wrong

Reply:
Would it be better to just go ahead and add `Item.RENAME` if we are changing the permission check anyway?